File: source/python/bubble.py
# bubble.py
import sys
import time

if len(sys.argv) != 2:
    print("usage: bubble.py <datafile>", file=sys.stderr)
    sys.exit(1)

with open(sys.argv[1]) as f:
    arr = [int(line) for line in f]

n = len(arr)

start = time.monotonic()

for i in range(n - 1):
    for j in range(n - 1 - i):
        if arr[j] > arr[j + 1]:
            arr[j], arr[j + 1] = arr[j + 1], arr[j]

elapsed = time.monotonic() - start
print(f"n={n} time={elapsed:.3f}s")

File: source/c/bubble.c
// bubble.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "usage: bubble <datafile>\n");
        return 1;
    }

    FILE *fp = fopen(argv[1], "r");
    if (!fp) {
        perror("fopen");
        return 1;
    }

    int cap = 50000;
    int *arr = malloc(cap * sizeof(int));
    int n = 0;

    while (fscanf(fp, "%d", &arr[n]) == 1)
        n++;
    fclose(fp);

    struct timespec start, end;
    clock_gettime(CLOCK_MONOTONIC, &start);

    for (int i = 0; i < n - 1; i++)
        for (int j = 0; j < n - 1 - i; j++)
            if (arr[j] > arr[j + 1]) {
                int tmp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = tmp;
            }

    clock_gettime(CLOCK_MONOTONIC, &end);

    double elapsed = (end.tv_sec - start.tv_sec) + (end.tv_nsec - start.tv_nsec) / 1e9;
    printf("n=%d time=%.3fs\n", n, elapsed);

    free(arr);
    return 0;
}
